File: app/security.py
import base64
import hashlib
import os

from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from itsdangerous import BadSignature, URLSafeTimedSerializer

from app.config import settings

_hasher = PasswordHasher()
_serializer = URLSafeTimedSerializer(settings.secret_key, salt="keyshort-session")
#passwords / passphrases 

def hash_secret(raw: str) -> str:
    return _hasher.hash(raw)


def verify_secret(hashed: str, raw: str) -> bool:
    try:
        return _hasher.verify(hashed, raw)
    except (VerifyMismatchError, Exception):  # noqa: BLE001 - any argon2 failure means "no"
        return False


# ---------- sessions ----------

def make_session(user_id: int) -> str:
    return _serializer.dumps({"uid": user_id})


def read_session(token: str) -> int | None:
    try:
        data = _serializer.loads(token, max_age=settings.session_max_age)
    except BadSignature:
        return None
    except Exception:  # noqa: BLE001
        return None
    uid = data.get("uid")
    return int(uid) if uid is not None else None


# ---------- provider key encryption (AES-256-GCM) ----------

def _aes_key() -> bytes:
    raw = settings.encryption_key.strip()
    if not raw:
        # Deterministic dev fallback so the app boots; production must set ENCRYPTION_KEY.
        return hashlib.sha256(settings.secret_key.encode()).digest()
    key = base64.b64decode(raw)
    if len(key) != 32:
        raise ValueError("ENCRYPTION_KEY must be base64 of exactly 32 bytes")
    return key


def encrypt_secret(plaintext: str) -> str:
    nonce = os.urandom(12)
    ct = AESGCM(_aes_key()).encrypt(nonce, plaintext.encode(), None)
    return base64.b64encode(nonce + ct).decode()


def decrypt_secret(blob: str) -> str:
    raw = base64.b64decode(blob)
    nonce, ct = raw[:12], raw[12:]
    return AESGCM(_aes_key()).decrypt(nonce, ct, None).decode()


def generate_encryption_key() -> str:
    """Helper for the README: prints a valid ENCRYPTION_KEY."""
    return base64.b64encode(os.urandom(32)).decode()


if __name__ == "__main__":
    print(generate_encryption_key())

File: src/security.js
const crypto = require("crypto");
const argon2 = require("argon2");
const { settings } = require("./config");

// ---------- passwords / passphrases ----------

async function hashSecret(raw) {
  return argon2.hash(raw, { type: argon2.argon2id });
}

async function verifySecret(hashed, raw) {
  if (!hashed) return false;
  try {
    return await argon2.verify(hashed, raw);
  } catch (err) {
    // any argon2 failure (malformed hash, mismatch) means "no"
    return false;
  }
}

// ---------- sessions ----------
//
// A small hand-rolled equivalent of itsdangerous's URLSafeTimedSerializer:
// base64url(JSON payload) + "." + base64url(HMAC-SHA256 signature).
// The payload carries an `iat` (issued-at) timestamp so we can enforce
// settings.sessionMaxAge the same way the Python side used `max_age`.

function base64url(buf) {
  return buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function fromBase64url(str) {
  let padded = str.replace(/-/g, "+").replace(/_/g, "/");
  while (padded.length % 4) padded += "=";
  return Buffer.from(padded, "base64");
}

function sign(payloadB64) {
  return base64url(crypto.createHmac("sha256", settings.secretKey).update(payloadB64).digest());
}

function makeSession(userId) {
  const payload = { uid: userId, iat: Math.floor(Date.now() / 1000) };
  const payloadB64 = base64url(Buffer.from(JSON.stringify(payload)));
  return `${payloadB64}.${sign(payloadB64)}`;
}

function readSession(token) {
  if (!token || typeof token !== "string") return null;
  try {
    const [payloadB64, sigB64] = token.split(".");
    if (!payloadB64 || !sigB64) return null;

    const expected = sign(payloadB64);
    const a = Buffer.from(sigB64);
    const b = Buffer.from(expected);
    if (a.length !== b.length || !crypto.timingSafeEqual(a, b)) return null;

    const payload = JSON.parse(fromBase64url(payloadB64).toString("utf8"));
    if (typeof payload.iat !== "number") return null;

    const age = Math.floor(Date.now() / 1000) - payload.iat;
    if (age > settings.sessionMaxAge) return null;

    return payload.uid !== undefined && payload.uid !== null ? Number(payload.uid) : null;
  } catch (err) {
    return null;
  }
}

// ---------- provider key encryption (AES-256-GCM) ----------

function aesKey() {
  const raw = (settings.encryptionKey || "").trim();
  if (!raw) {
    // Deterministic dev fallback so the app boots; production must set ENCRYPTION_KEY.
    return crypto.createHash("sha256").update(settings.secretKey).digest();
  }
  const key = Buffer.from(raw, "base64");
  if (key.length !== 32) {
    throw new Error("ENCRYPTION_KEY must be base64 of exactly 32 bytes");
  }
  return key;
}

function encryptSecret(plaintext) {
  const nonce = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", aesKey(), nonce);
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  // Layout matches Python's `cryptography` AESGCM: nonce + ciphertext + tag.
  return Buffer.concat([nonce, ciphertext, tag]).toString("base64");
}

function decryptSecret(blob) {
  const raw = Buffer.from(blob, "base64");
  const nonce = raw.subarray(0, 12);
  const tag = raw.subarray(raw.length - 16);
  const ciphertext = raw.subarray(12, raw.length - 16);
  const decipher = crypto.createDecipheriv("aes-256-gcm", aesKey(), nonce);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf8");
}

/** Helper for the README: prints a valid ENCRYPTION_KEY. */
function generateEncryptionKey() {
  return crypto.randomBytes(32).toString("base64");
}

module.exports = {
  hashSecret,
  verifySecret,
  makeSession,
  readSession,
  encryptSecret,
  decryptSecret,
  generateEncryptionKey,
};
